pulp_container/tests/functional/api/test_push_content.py

```diff
@@ -4,52 +4,140 @@
 
-from pulpcore.client.pulp_container import RepositoriesContainerPushApi
+from pulpcore.client.pulp_container import DistributionsContainerApi, RepositoriesContainerPushApi
 
 
-cfg = config.get_config()
+class PushRepoTestCase(unittest.TestCase):
+    """Verify whether images can be pushed to pulp."""
+    @classmethod
+    def setUpClass(cls):
+        """
+        Define APIs to use and pull images needed later in tests
+        """
+        api_client = gen_container_client()
+        cls.distribution_api = DistributionsContainerApi(api_client)
+        cls.pushrepository_api = RepositoriesContainerPushApi(api_client)
 
-api_client = gen_container_client()
-push_repositories_api = RepositoriesContainerPushApi(api_client)
+        cls.cfg = config.get_config()
+        cls.registry = cli.RegistryClient(cls.cfg)
+        cls.registry.raise_if_unsupported(unittest.SkipTest, "Tests require podman/docker")
+        # TODO better handling of the "http://"
+        cls.registry_name = cls.cfg.get_base_url()[7:]
 
-registry = cli.RegistryClient(cfg)
+        cls.user_creator = gen_user([
+            "container.add_containerdistribution",
+            "container.view_containerdistribution",
+            "container.add_containernamespace"
+        ])
+        cls.user_reader = gen_user(["container.view_containerpushrepository"])
+        cls.user_helpless = gen_user([])
 
+        cls._pull("centos:7")
+        cls._pull(f"{DOCKERHUB_PULP_FIXTURE_1}:manifest_a")
 
-class PushContentTestCase(unittest.TestCase):
-    """Verify whether images can be pushed to pulp."""
+    @classmethod
+    def tearDownClass(cls):
+        """Delete api users."""
+        del_user(cls.user_creator)
+        del_user(cls.user_reader)
+        del_user(cls.user_helpless)
 
-    def test_push_using_registry_client(self):
-        """Test push with official registry client"""
-        registry.raise_if_unsupported(unittest.SkipTest, "Test requires podman/docker")
-        # TODO better handling of the "http://"
-        registry_name = cfg.get_base_url()[7:]
-        local_url = "/".join([registry_name, "foo/bar:1.0"])
-        # Be sure to not being logged in
-        try:
-            registry.logout(registry_name)
-        except exceptions.CalledProcessError:
-            pass
-        # Pull an image with large blobs
-        registry.pull("centos:7")
+    @classmethod
+    def _pull(cls, image_path, user=None):
+        """
+        Pull using specified user.
+        """
+        if user:
+            cls.registry.login("-u", user['username'], "-p", user['password'], cls.registry_name)
+        else:
+            # Ensure logout
+            try:
+                cls.registry.logout(cls.registry_name)
+            except exceptions.CalledProcessError:
+                pass
+
+        cls.registry.pull(image_path)
+
+        if user:
+            cls.registry.logout(cls.registry_name)
+
+    @classmethod
+    def _push(cls, image_path, local_url, user):
```

Comment:
This function is performing much more than a simple push atm. It makes reading the tests very confusing.

Reply:
I can copy-paste it into every test or rename to tag and push.
It's worsens readability of the test significantly (for me personally) to have all those lines of code in every test just for the sake of making push while being logged in.

I'm also thinking that I might just drop all the tests when they work anyway, since the private/public repo work will change those entirely.